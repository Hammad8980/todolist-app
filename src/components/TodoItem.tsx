export type Task = {
  id: number;
  title: string;
  isCompleted: boolean;
  priority?: 'p1' | 'p2' | 'p3';
};
