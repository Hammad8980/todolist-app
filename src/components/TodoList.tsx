import type { Task } from './TodoItem'
type TodoListProps = {
  tasks : Task[];
}
export default function TodoList({tasks}: TodoListProps) {
  return(
    <ul>
          {tasks.map((tasks) => (
            <li key={tasks.id}>{tasks.title}</li>
          ))}
    </ul>
  )

}

;
