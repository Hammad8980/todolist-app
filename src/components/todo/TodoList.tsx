import { type PropsWithChildren } from 'react';
import { type Task } from '../../features/todos/TodoTaskTypes';

type TodoItemProps = PropsWithChildren<{
  task: Task;
  onDelete: (id: number) => void;
  onToggle: (id: number) => void;
}>;

export default function TodoItem({ task, onDelete, onToggle, children }: TodoItemProps) {
  return (
    <li className="flex items-center gap-2 py-2 border-b">
      <input 
        type="checkbox" 
        checked={task.isCompleted}
        onChange={() => onToggle(task.id)}
        className="h-5 w-5"
      />
      <span className={task.isCompleted ? "line-through text-gray-500" : ""}>
        {task.title}
      </span>
      <button 
        onClick={() => onDelete(task.id)}
        className="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        Delete
      </button>
      <span className="text-sm text-gray-500  border-l pl-2">
        {task.priority === 'p1' ? (
          <span className="text-green-500">Low</span>
        ) : task.priority === 'p2' ? (
          <span className="text-yellow-500">Medium</span>
        ) : task.priority === 'p3' ? (
          <span className="text-red-500">High</span>
        ) : (
          <span className="text-gray-500">Unknown</span>
        )}
      </span>
      {children}
    </li>
  );
}