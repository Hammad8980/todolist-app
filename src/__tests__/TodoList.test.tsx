import { render, screen } from '@testing-library/react';
import TodoList from '../components/todo/TodoList';
import TodoItem from '../components/todo/TodoItem';
import type { Task } from '../features/todos/TodoTaskTypes';

describe('TodoList', () => {
  const mockOnDelete = jest.fn();
  const mockOnToggle = jest.fn();

  const sampleTasks: Task[] = [
    {
      id: 1,
      title: 'First task',
      isCompleted: false,
      priority: 'p1',
    },
    {
      id: 2,
      title: 'Second task',
      isCompleted: true,
      priority: 'p2',
    },
    {
      id: 3,
      title: 'Third task',
      isCompleted: false,
      priority: 'p3',
    },
  ];

  beforeEach(() => {
    mockOnDelete.mockClear();
    mockOnToggle.mockClear();
  });
  // For the TodoItem component, we can mock it to simplify the tests
  it('Mocks TodoItem component with sample task', () => {
    const mockTask: Task = {
      id: 1,
      title: 'Mock task',
      isCompleted: false,
      priority: 'p1',
    };

    const { container } = render(
      <ul>
        <TodoItem task={mockTask} onDelete={mockOnDelete} onToggle={mockOnToggle} />
      </ul>
    );
    expect(screen.getByText('Mock task')).toBeInTheDocument();
    expect(screen.getByRole('checkbox')).not.toBeChecked();
    expect(screen.getByRole('button', { name: 'Delete' })).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  // Below are the tests for the TodoList component
  it('renders empty state when no tasks are provided', () => {
    const { container } = render(<TodoList tasks={[]} onDelete={mockOnDelete} onToggle={mockOnToggle} />);

    expect(screen.getByText(/no tasks yet/i)).toBeInTheDocument();
    expect(screen.getByText(/add your first task above/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('renders all tasks when tasks are provided', () => {
    const { container } = render(<TodoList tasks={sampleTasks} onDelete={mockOnDelete} onToggle={mockOnToggle} />);

    expect(screen.getByText('First task')).toBeInTheDocument();
    expect(screen.getByText('Second task')).toBeInTheDocument();
    expect(screen.getByText('Third task')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('renders the correct number of tasks', () => {
    const { container } = render(<TodoList tasks={sampleTasks} onDelete={mockOnDelete} onToggle={mockOnToggle} />);

    const listItems = screen.getAllByRole('listitem');
    expect(listItems).toHaveLength(3);
    expect(container).toMatchSnapshot();
  });

  it('passes correct props to TodoItem components', () => {
    const { container } = render(<TodoList tasks={sampleTasks} onDelete={mockOnDelete} onToggle={mockOnToggle} />);

    // Check that checkboxes are rendered with correct states
    const checkboxes = screen.getAllByRole('checkbox');
    expect(checkboxes).toHaveLength(3);
    expect(checkboxes[0]).not.toBeChecked(); // First task
    expect(checkboxes[1]).toBeChecked(); // Second task (completed)
    expect(checkboxes[2]).not.toBeChecked(); // Third task
    expect(container).toMatchSnapshot();
  });
});
