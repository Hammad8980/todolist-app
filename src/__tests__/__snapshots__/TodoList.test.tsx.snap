// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Mocks TodoItem component with sample task 1`] = `
<div>
  <ul>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        Mock task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
  </ul>
</div>
`;

exports[`TodoList passes correct props to TodoItem components 1`] = `
<div>
  <ul
    class="space-y-2"
  >
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        First task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        checked=""
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class="line-through text-gray-500"
      >
        Second task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        Third task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
  </ul>
</div>
`;

exports[`TodoList renders all tasks when tasks are provided 1`] = `
<div>
  <ul
    class="space-y-2"
  >
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        First task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        checked=""
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class="line-through text-gray-500"
      >
        Second task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        Third task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
  </ul>
</div>
`;

exports[`TodoList renders empty state when no tasks are provided 1`] = `
<div>
  <div
    class="text-center py-8"
  >
    <p
      class="text-gray-500"
    >
      No tasks yet! Add your first task above.
    </p>
  </div>
</div>
`;

exports[`TodoList renders the correct number of tasks 1`] = `
<div>
  <ul
    class="space-y-2"
  >
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        First task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        checked=""
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class="line-through text-gray-500"
      >
        Second task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        Third task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
  </ul>
</div>
`;
